flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicCommitter.java

```diff
@@ -0,0 +1,420 @@
+    if (commitRequests.isEmpty()) {
+      return;
+    }
+
+    Map<TableKey, NavigableMap<Long, List<CommitRequest<DynamicCommittable>>>> commitRequestMap =
+        Maps.newHashMap();
+    for (CommitRequest<DynamicCommittable> request : commitRequests) {
+      NavigableMap<Long, List<CommitRequest<DynamicCommittable>>> committables =
+          commitRequestMap.computeIfAbsent(
+              new TableKey(request.getCommittable()), unused -> Maps.newTreeMap());
+      committables
+          .computeIfAbsent(request.getCommittable().checkpointId(), unused -> Lists.newArrayList())
+          .add(request);
+    }
+
+    for (Map.Entry<TableKey, NavigableMap<Long, List<CommitRequest<DynamicCommittable>>>> entry :
+        commitRequestMap.entrySet()) {
+      Table table = catalog.loadTable(TableIdentifier.parse(entry.getKey().tableName()));
+      DynamicCommittable last = entry.getValue().lastEntry().getValue().get(0).getCommittable();
+      long maxCommittedCheckpointId =
+          getMaxCommittedCheckpointId(
+              table, last.jobId(), last.operatorId(), entry.getKey().branch());
+      // Mark the already committed FilesCommittable(s) as finished
+      entry
+          .getValue()
+          .headMap(maxCommittedCheckpointId, true)
+          .values()
+          .forEach(list -> list.forEach(CommitRequest::signalAlreadyCommitted));
+      NavigableMap<Long, List<CommitRequest<DynamicCommittable>>> uncommitted =
+          entry.getValue().tailMap(maxCommittedCheckpointId, false);
+      if (!uncommitted.isEmpty()) {
+        commitPendingRequests(
+            table, entry.getKey().branch(), uncommitted, last.jobId(), last.operatorId());
+      }
+    }
+  }
+
+  private static long getMaxCommittedCheckpointId(
+      Table table, String flinkJobId, String operatorId, String branch) {
+    Snapshot snapshot = table.snapshot(branch);
+    long lastCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
+
+    while (snapshot != null) {
+      Map<String, String> summary = snapshot.summary();
+      String snapshotFlinkJobId = summary.get(FLINK_JOB_ID);
+      String snapshotOperatorId = summary.get(OPERATOR_ID);
+      if (flinkJobId.equals(snapshotFlinkJobId)
+          && (snapshotOperatorId == null || snapshotOperatorId.equals(operatorId))) {
+        String value = summary.get(MAX_COMMITTED_CHECKPOINT_ID);
+        if (value != null) {
+          lastCommittedCheckpointId = Long.parseLong(value);
+          break;
+        }
+      }
+
+      Long parentSnapshotId = snapshot.parentId();
+      snapshot = parentSnapshotId != null ? table.snapshot(parentSnapshotId) : null;
+    }
+
+    return lastCommittedCheckpointId;
+  }
+
+  /**
+   * Commits the data to the Iceberg table by reading the file data from the {@link DeltaManifests}
+   * ordered by the checkpointId, and writing the new snapshot to the Iceberg table. The {@link
+   * org.apache.iceberg.SnapshotSummary} will contain the jobId, snapshotId, checkpointId so in case
```

Comment:
nit: not that important, but we might want to import `org.apache.iceberg.SnapshotSummary` so the javadoc is shorter?

Really only do it if you touch this part